### PR TITLE
Mobile batch results

### LINKED ISSUE #348: Hamburger menu not rendered on top (mobile view)
See screenshot below.

image

### PR DESCRIPTION
**Pull request by binh-dam-ibigroup**
This PR displays the list of itinerary options in mobile views when batch routing is enabled.

This PR builds upon #337. Behind the scenes, the code that renders headers and error screen is shared between the existing and new mobile result view and extracted into ~its~ their own component**s**.

Also: A new `ui_itineraryView` URL parameter is added to reflect the itinerary expanded state, right now it is available only for mobile batch results.

EDIT: Also fix #348.

Screenshot:
image

### CODE REVIEW

**Comment by binh-dam-ibigroup** [comment]
I marked this PR as blocked pending #337.

File: example.js
Comment (evansiroky) [inline example.js:76]: Add this component as a required component in the comment about the creation of the `components` variable.

**Review by evansiroky — APPROVED** [review]
Approved pending resolution of 1 comment

File: example.js
Comment (binh-dam-ibigroup) [reply example.js:77]: Updated in https://github.com/opentripplanner/otp-react-redux/pull/339/commits/e802be61940f6d6454d6c6291b61e310bdf9ae21.

**Review by landonreed — CHANGES_REQUESTED** [review]
I think a few things might need some work, but this is headed in the right direction!

File: lib/components/mobile/batch-results-screen.js
Comment (binh-dam-ibigroup) [reply lib/components/mobile/batch-results-screen.js:87]: This is what I see in the body if there is an error. What do you see?

image

File: lib/components/mobile/batch-results-screen.js
Comment (binh-dam-ibigroup) [reply lib/components/mobile/batch-results-screen.js]: Updated in https://github.com/opentripplanner/otp-react-redux/pull/339/commits/100c1d62160bc53210f50732875aa911ec4ed982.

File: lib/components/mobile/batch-results-screen.js
Comment (binh-dam-ibigroup) [reply lib/components/mobile/batch-results-screen.js]: Updated in https://github.com/opentripplanner/otp-react-redux/pull/339/commits/100c1d62160bc53210f50732875aa911ec4ed982.

File: lib/components/mobile/batch-results-screen.js
Comment (binh-dam-ibigroup) [reply lib/components/mobile/batch-results-screen.js:54]: Yes, I added a comment regarding that in https://github.com/opentripplanner/otp-react-redux/pull/339/commits/100c1d62160bc53210f50732875aa911ec4ed982.

Comment (binh-dam-ibigroup) [reply lib/components/mobile/batch-results-screen.js:54]: Also, I merged the commits from #342, so you should see the effect from clicking a different leg now.

File: lib/components/mobile/batch-results-screen.js
Comment (binh-dam-ibigroup) [reply lib/components/mobile/batch-results-screen.js:40]: Updated in https://github.com/opentripplanner/otp-react-redux/pull/339/commits/100c1d62160bc53210f50732875aa911ec4ed982.

File: lib/components/mobile/batch-results-screen.js
Comment (binh-dam-ibigroup) [reply lib/components/mobile/batch-results-screen.js:92]: The "impossible" itinerary you tried, the previous behavior does not change screens as a result of clicking the Plan button, so I'll have to investigate more.

File: lib/components/mobile/results-header-and-error.js
Comment (binh-dam-ibigroup) [reply lib/components/mobile/results-header-and-error.js]: Updated in https://github.com/opentripplanner/otp-react-redux/pull/339/commits/100c1d62160bc53210f50732875aa911ec4ed982.

File: lib/components/mobile/results-header-and-error.js
Comment (binh-dam-ibigroup) [reply lib/components/mobile/results-header-and-error.js]: Addressed in https://github.com/opentripplanner/otp-react-redux/pull/339/commits/df664e04cae78655b5e26907567b2b81004ebd72.

File: lib/components/mobile/batch-results-screen.js
Comment (binh-dam-ibigroup) [reply lib/components/mobile/batch-results-screen.js]: The behavior that was previously missing for "impossible" itineraries (including the old mobile view) is addressed in https://github.com/opentripplanner/otp-react-redux/pull/339/commits/79488a8755802c13712df91e0b45a316c79dc7ff.

File: lib/components/mobile/batch-results-screen.js
Comment (binh-dam-ibigroup) [reply lib/components/mobile/batch-results-screen.js:95]: If we don't rerender the map and simply resize it (what we'd have to do to enable the animated transition), then the inner map canvas is not properly resized... any suggestions for triggering that?
image

Comment (binh-dam-ibigroup) [reply lib/components/mobile/batch-results-screen.js:95]: I think you would have to reload the tiles anyway after expanding/collapsing the map because the zoom and fit would be different in either case.

Comment (binh-dam-ibigroup) [reply lib/components/mobile/batch-results-screen.js]: Okay, I found a way to do that with https://github.com/opentripplanner/otp-react-redux/pull/339/commits/557bfc7387d771ce1a21c3c36fe7e0f65a1e28c4 (after almost giving up... - see the transitive trick in https://github.com/opentripplanner/otp-react-redux/pull/339/commits/2e7757c9365880a36f3bb154f1beaa1fb3d16556)

File: lib/components/mobile/batch-results-screen.js
Comment (binh-dam-ibigroup) [reply lib/components/mobile/batch-results-screen.js]: I went ahead and added a URL parameter `ui_itineraryView` in https://github.com/opentripplanner/otp-react-redux/pull/339/commits/3edc3d4f6867f3c6a3d18c11cd114dd9334ca44b to replace passing props around to track the map expanded state.

**Comment by binh-dam-ibigroup** [comment]
Ready for review again. The only remaining issue I have not addressed is whether to expand the itinerary pane and collapse the map if user clicks again on the same active leg. Otherwise I'm open to renaming/moving things around.

File: lib/components/map/map.js
Comment (evansiroky) [inline lib/components/map/map.js:17]: Why this change? Since there is no access to instance variables, I don't think this is needed.

File: lib/actions/ui.js
Comment (evansiroky) [inline lib/actions/ui.js:242]: Please add JSDoc explaining what these represent.

File: lib/components/mobile/results-screen.js
Comment (evansiroky) [inline lib/components/mobile/results-screen.js:118]: Should this be made into a button element since it's clickable?

File: lib/components/mobile/results-screen.js
Comment (evansiroky) [inline lib/components/mobile/results-screen.js:119]: This could be part of a styled component that accepts props and styles accordingly.

File: lib/components/mobile/results-screen.js
Comment (evansiroky) [inline lib/components/mobile/results-screen.js:122]: This could be part of a styled component that accepts props and styles accordingly.

File: lib/components/mobile/batch-results-screen.js
Comment (evansiroky) [inline lib/components/mobile/batch-results-screen.js:82]: This could be part of a styled component that accepts props and styles accordingly.

File: lib/components/mobile/results-error.js
Comment (evansiroky) [inline lib/components/mobile/results-error.js:28]: If instead a `className` prop is passed, this higher-order component can be styled as a styled component. [See these docs](https://styled-components.com/docs/advanced#caveat).

File: lib/components/mobile/batch-results-screen.js
Comment (evansiroky) [inline lib/components/mobile/batch-results-screen.js:100]: If ResultsError is changed to be styled-components-compatible (see https://github.com/opentripplanner/otp-react-redux/pull/339/files#r602670869), then this ResultsError component could be wrapped as a styled component that accepts props and styles accordingly. [See these docs](https://styled-components.com/docs/basics#adapting-based-on-props).

**Review by evansiroky — APPROVED** [review]
The mobile UX is now pretty great! I have a few ideas for improvement but seems good to go other than those things.

File: lib/components/mobile/results-header.js
Comment (landonreed) [inline lib/components/mobile/results-header.js:59]: In batch mode, if you have planned a trip and expanded an itinerary, clicking this takes you back to the search form, but when you re-plan the trip it takes you directly to the expanded itinerary view. Instead, it should take you back to the results list.

File: lib/actions/ui.js
Comment (landonreed) [inline lib/actions/ui.js:246]: This is for mobile only, right? I think that should be made clear here.

File: lib/components/map/bounds-updating-overlay.js
Comment (landonreed) [inline lib/components/map/bounds-updating-overlay.js:76]: This is perhaps the wrong place for this comment, but when you click a leg summary to show the split screen map view, is there any way to go back to the full itinerary view?

File: lib/components/map/bounds-updating-overlay.js
Comment (landonreed) [inline lib/components/map/bounds-updating-overlay.js:79]:
```suggestion
      // and do that after a delay to ensure that canvas heights have stabilized in the DOM.
```

File: lib/components/map/bounds-updating-overlay.js
Comment (landonreed) [inline lib/components/map/bounds-updating-overlay.js:95]: Non-blocking: This whole block might be nice in a `fitMapToItinerary` method.

File: lib/components/narrative/narrative-itineraries.js
Comment (landonreed) [inline lib/components/narrative/narrative-itineraries.js:52]: This isn't the right spot for this comment, but I think `NarrativeItinerary#_onHeaderClick` might need to invoke the `setItineraryView` action:

https://github.com/opentripplanner/otp-react-redux/blob/ef9554794272cbac5b73c75772e39dd69e674315/lib/components/narrative/narrative-itinerary.js#L20-L29

**Review by landonreed — CHANGES_REQUESTED** [review]
To echo what @evansiroky said, this is a proper mobile UI! Nice work! Just a few changes to get it over the finish line.

File: lib/components/map/map.js
Comment (binh-dam-ibigroup) [reply lib/components/map/map.js]: Reverted in https://github.com/opentripplanner/otp-react-redux/pull/339/commits/794aeafcdf59f7ea394a00aac565daa4e7a9b2f0.

File: lib/actions/ui.js
Comment (binh-dam-ibigroup) [reply lib/actions/ui.js:246]: I added the JSDoc in https://github.com/opentripplanner/otp-react-redux/pull/339/commits/794aeafcdf59f7ea394a00aac565daa4e7a9b2f0.

File: lib/components/map/bounds-updating-overlay.js
Comment (binh-dam-ibigroup) [reply lib/components/map/bounds-updating-overlay.js]: Updated in https://github.com/opentripplanner/otp-react-redux/pull/339/commits/794aeafcdf59f7ea394a00aac565daa4e7a9b2f0.

File: lib/components/map/bounds-updating-overlay.js
Comment (binh-dam-ibigroup) [reply lib/components/map/bounds-updating-overlay.js:98]: I understood that #133 was fixed in some way by #134... any insights in reproducing it?

Comment (binh-dam-ibigroup) [reply lib/components/map/bounds-updating-overlay.js:98]: Probably better to deal with it separately (https://github.com/opentripplanner/otp-react-redux/pull/339/commits/b1ab768e30ed011781d6ab4c4174a6ad746949d0)

File: lib/components/narrative/narrative-itineraries.js
Comment (binh-dam-ibigroup) [reply lib/components/narrative/narrative-itineraries.js:52]: Updated in https://github.com/opentripplanner/otp-react-redux/pull/339/commits/19f056929e4f8ee17fadc8f1ac3c795ee29f3878.

File: lib/components/map/bounds-updating-overlay.js
Comment (binh-dam-ibigroup) [reply lib/components/map/bounds-updating-overlay.js]: Updated in https://github.com/opentripplanner/otp-react-redux/pull/339/commits/b738bd20222e0614ab23025d94634b769981bac7.

File: lib/components/mobile/results-header.js
Comment (binh-dam-ibigroup) [reply lib/components/mobile/results-header.js:59]: Good catch, fixed in https://github.com/opentripplanner/otp-react-redux/pull/339/commits/df927a8f852ee3ec599d8f14bdd44406cf9376af.

File: lib/components/map/bounds-updating-overlay.js
Comment (binh-dam-ibigroup) [reply lib/components/map/bounds-updating-overlay.js:100]: Good idea to insist. In https://github.com/opentripplanner/otp-react-redux/pull/339/commits/7678132763b027dc21ff3d7784230b7fe5577913, clicking on the same leg in the narrative will toggle between the split map and the full itinerary view.

File: lib/components/mobile/results-screen.js
Comment (binh-dam-ibigroup) [reply lib/components/mobile/results-screen.js]: Updated in https://github.com/opentripplanner/otp-react-redux/pull/339/commits/77bfc8f678830e18fa9ed828d9d6e6652969747a

File: lib/components/mobile/batch-results-screen.js
Comment (binh-dam-ibigroup) [reply lib/components/mobile/batch-results-screen.js:78]: Generally, existing CSS doesn't play nice with styled components. Making this change involves plucking the existing CSS apart, and I would stay away from that until further discussion.

Comment (binh-dam-ibigroup) [reply lib/components/mobile/batch-results-screen.js]: Alright, this is converted in https://github.com/opentripplanner/otp-react-redux/pull/339/commits/d7737831bffd7e6a4f5294541b0187c1a2574c3a.

File: lib/components/mobile/results-error.js
Comment (binh-dam-ibigroup) [reply lib/components/mobile/results-error.js]: Updated in https://github.com/opentripplanner/otp-react-redux/pull/339/commits/d7737831bffd7e6a4f5294541b0187c1a2574c3a.

File: lib/components/mobile/batch-results-screen.js
Comment (binh-dam-ibigroup) [reply lib/components/mobile/batch-results-screen.js]: Updated in https://github.com/opentripplanner/otp-react-redux/pull/339/commits/d7737831bffd7e6a4f5294541b0187c1a2574c3a

**Comment by binh-dam-ibigroup** [comment]
Ready for review again...

File: lib/components/mobile/batch-results-screen.js
Comment (landonreed) [inline lib/components/mobile/batch-results-screen.js:61]: On second look, I find the re-assignment/renaming of ItineraryView a bit confusing. It feels like a bit of an unnecessary redirection for the reader. Recommendation: `const {ItineraryView} = uiActions`

File: lib/components/mobile/batch-results-screen.js
Comment (landonreed) [inline lib/components/mobile/batch-results-screen.js:78]: long line, please split up. Also, I think this could use a comment. What is the behavior here? The use of `previousItineraryView` is kind of confusing to me, especially given that the setState call happens within `_setItineraryView`.

File: lib/components/narrative/narrative-itineraries.js
Comment (landonreed) [inline lib/components/narrative/narrative-itineraries.js:15]: I suppose this file isn't already doing this, but I do prefer these action imports to follow the format used elsewhere, i.e.:

```
import * as uiActions from '../../actions/ui'
```

**Review by landonreed — CHANGES_REQUESTED** [review]
Generally looks pretty good! Just a couple of minor items I flagged. Mainly I just find the use of `this.state.previousItineraryView` kind of confusing. I wonder if there's a better way?

File: lib/components/narrative/narrative-itineraries.js
Comment (binh-dam-ibigroup) [reply lib/components/narrative/narrative-itineraries.js]: Fixed in https://github.com/opentripplanner/otp-react-redux/pull/339/commits/48bafb663209bf2e84d22b0c491709cba7c31919.

File: lib/components/mobile/batch-results-screen.js
Comment (binh-dam-ibigroup) [reply lib/components/mobile/batch-results-screen.js]: Updated in https://github.com/opentripplanner/otp-react-redux/pull/339/commits/48bafb663209bf2e84d22b0c491709cba7c31919. We need the previous state to correctly display the itinerary as a list or the detailed itinerary when the user clicks `Show results`.

File: lib/components/mobile/batch-results-screen.js
Comment (binh-dam-ibigroup) [reply lib/components/mobile/batch-results-screen.js]: Updated in https://github.com/opentripplanner/otp-react-redux/pull/339/commits/48bafb663209bf2e84d22b0c491709cba7c31919.

File: lib/actions/ui.js
Comment (evansiroky) [inline lib/actions/ui.js:247]: Can each of these have a comment about what they mean?

File: lib/components/mobile/batch-results-screen.js
Comment (evansiroky) [inline lib/components/mobile/batch-results-screen.js:98]: Can this variable be renamed to say that there are no itineraries and errors exist?

File: lib/components/narrative/narrative-itineraries.js
Comment (evansiroky) [inline lib/components/narrative/narrative-itineraries.js:230]: How many of the props are really needed here? In general, I think spreading props to child components should try to be avoided if possible.

**Review by evansiroky — CHANGES_REQUESTED** [review]
See comments. Also 2 usability problems:

1. Trip plans with errors are still showing up when a single itinerary is expanded. See picture:

Screen Shot 2021-04-01 at 2 53 29 PM

2. If I expand the map, then try to edit the trip parameters, then hit search, the itinerary options don't seem to show up again.

File: lib/components/narrative/narrative-itineraries.js
Comment (binh-dam-ibigroup) [reply lib/components/narrative/narrative-itineraries.js]: Addressed, hopefully, in https://github.com/opentripplanner/otp-react-redux/pull/339/commits/3b2b8801f7c708392ca787f0401dc08f6be9103b.

File: lib/components/mobile/batch-results-screen.js
Comment (binh-dam-ibigroup) [reply lib/components/mobile/batch-results-screen.js]: Updated in https://github.com/opentripplanner/otp-react-redux/pull/339/commits/cf678bbf51da21206f401c18e87cb719c805f6ff.

File: lib/actions/ui.js
Comment (binh-dam-ibigroup) [reply lib/actions/ui.js]: Updated in https://github.com/opentripplanner/otp-react-redux/pull/339/commits/cf678bbf51da21206f401c18e87cb719c805f6ff.

**Comment by binh-dam-ibigroup** [comment]
> See comments. Also 2 usability problems:
> 
> 1. Trip plans with errors are still showing up when a single itinerary is expanded. See picture:
> 
> Screen Shot 2021-04-01 at 2 53 29 PM
> 
> 1. If I expand the map, then try to edit the trip parameters, then hit search, the itinerary options don't seem to show up again.

1.  was fixed in https://github.com/opentripplanner/otp-react-redux/pull/339/commits/cf678bbf51da21206f401c18e87cb719c805f6ff.
2. I am no longer able to reproduce after https://github.com/opentripplanner/otp-react-redux/pull/339/commits/3b2b8801f7c708392ca787f0401dc08f6be9103b... can you?

**Comment by evansiroky** [comment]
> > 1. If I expand the map, then try to edit the trip parameters, then hit search, the itinerary options don't seem to show up again.
> 2. I am no longer able to reproduce after [3b2b880](https://github.com/opentripplanner/otp-react-redux/commit/3b2b8801f7c708392ca787f0401dc08f6be9103b)... can you?

Yep, still happening for me. See GIF:

Screen Recording 2021-04-02 at 6 06 27 PM

**Review by evansiroky — CHANGES_REQUESTED** [review]
See comment

**Comment by binh-dam-ibigroup** [comment]
Thanks for the steps @evansiroky, I found the needle in the haystack (https://github.com/opentripplanner/otp-react-redux/pull/339/commits/5d6a39383a7fdf48555477b87a800237921fa56f).

**Comment by landonreed** [comment]
@binh-dam-ibigroup I'm not sure if this is from recent changes, but when viewing a leg on the map and then clicking "Expand map" it momentarily flashes the list of options:


https://user-images.githubusercontent.com/2370911/114201897-afaaa280-9924-11eb-8abc-c935e1771a22.mov

**Comment by binh-dam-ibigroup** [comment]
> @binh-dam-ibigroup I'm not sure if this is from recent changes, but when viewing a leg on the map and then clicking "Expand map" it momentarily flashes the list of options:
> 
>  Screen.Recording.2021-04-09.at.11.12.03.AM.mov

Ugh side effect. Fixed in https://github.com/opentripplanner/otp-react-redux/pull/339/commits/af9f653f1d0dec934913ef1ec57fb42eeb995f42.

File: lib/components/narrative/narrative-itineraries.js
Comment (landonreed) [inline lib/components/narrative/narrative-itineraries.js:282]: I'm not sure if this is needed any longer?

**Review by landonreed — APPROVED** [review]
Looks pretty good! Just one comment on the `filter` prop. Let's wait for @evansiroky's review to merge on Monday.

File: lib/components/narrative/narrative-itineraries.js
Comment (binh-dam-ibigroup) [reply lib/components/narrative/narrative-itineraries.js]: Thanks, updated in https://github.com/opentripplanner/otp-react-redux/pull/339/commits/3f138bb57c187373a0dea19cc90aea90f8fb92bd.

File: lib/components/mobile/batch-results-screen.js
Comment (evansiroky) [inline lib/components/mobile/batch-results-screen.js:78]: Since this state property is always used in conjunction with a prop from the overall store, it might be cleaner to move this into the overall application state and also make the `toggleMapExpanded` handler be an action. That way, this component can become a purely presentational component. This recommendation is non-blocking.

File: lib/components/mobile/edit-search-button.js
Comment (evansiroky) [inline lib/components/mobile/edit-search-button.js:28]: It would simplify this component a fair amount if this method were converted into a connected action. This comment in non-blocking.

**Review by evansiroky — APPROVED** [review]
Just two non-blocking recommendations for improvement.

File: lib/components/mobile/edit-search-button.js
Comment (binh-dam-ibigroup) [reply lib/components/mobile/edit-search-button.js]: Thanks for the suggestion, updated in https://github.com/opentripplanner/otp-react-redux/pull/339/commits/43fef77367b6d788c401198500108f49fff1f355.

File: lib/components/mobile/batch-results-screen.js
Comment (binh-dam-ibigroup) [reply lib/components/mobile/batch-results-screen.js]: Thanks, this simplies the UI component a good bit (https://github.com/opentripplanner/otp-react-redux/pull/339/commits/43fef77367b6d788c401198500108f49fff1f355).

**Pull request merged**
